executor/hash_table.go

```diff
@@ -189,10 +237,50 @@ func (*hashRowContainer) getJoinKeyFromChkRow(sc *stmtctx.StatementContext, row
 	return false, hCtx.hashVals[0].Sum64(), err
 }
 
+// Len returns the length of the records in hashRowContainer.
 func (c hashRowContainer) Len() int {
 	return c.hashTable.Len()
 }
 
+func (c *hashRowContainer) Close() error {
+	if c.recordsInDisk != nil {
+		return c.recordsInDisk.Close()
+	}
+	return nil
+}
+
+// GetMemTracker returns the underlying memory usage tracker in hashRowContainer.
+func (c *hashRowContainer) GetMemTracker() *memory.Tracker { return c.memTracker }
+
+// GetDiskTracker returns the underlying disk usage tracker in hashRowContainer.
+func (c *hashRowContainer) GetDiskTracker() *disk.Tracker { return c.recordsInDisk.GetDiskTracker() }
+
+// SetSpillToDisk sets the memQuota for disk spilling, if records in
+// hashRowContainer exceeds the memQuota, All records and further chunks
+// putting by PutChunk will be stored to disk.
+func (c *hashRowContainer) SetSpillToDisk(memQuota int64) {
+	c.memTracker.SetBytesLimit(memQuota)
+	c.memTracker.SetActionOnExceed(&spillDiskAction{c: c})
+}
+
+// spillDiskAction implements memory.ActionOnExceed for chunk.List if
+// the memory quota of chunk.List is exceeded, spillDiskAction.Action
+// is triggered.
+type spillDiskAction struct {
+	once sync.Once
+	c    *hashRowContainer
+}
+
+// Action set the exceeded flag in hashRowContainer to true.
+func (a *spillDiskAction) Action(t *memory.Tracker) {
+	a.once.Do(func() {
+		a.c.exceeded = true
+		logutil.BgLogger().Info("memory exceeds quota, spill to disk now.", zap.String("memory", t.String()))
```

Comment:
How about adding the SQL(or the SQL fingerprint) to the log?

Reply:
Now we use the root `memory.Tracker`, problem solved.